nixos/modules/system/boot/resolved.nix

```diff
@@ -177,6 +221,14 @@ in
           assertion = !config.networking.useHostResolvConf;
           message = "Using host resolv.conf is not supported with systemd-resolved";
         }
+
+        {
+          assertion = !(cfg.mdns.enable && config.services.avahi.enable);
+          message = ''
+            mDNS resolution cannot be handled by both systemd-resolved and
+            Avahi at the same time.
+          '';
+        }
```

Comment:
I think this assertion will trip on a lot of existing configs. e.g. Enabling GNOME enables avahi by default. If they also have systemd-resolved enabled (which is common on its own, but also enabled from `networking.useNetworkd = true;`, which will hopefully be the default in 25.05), then mdns being enabled by default will trip this.

Reply:
Currently, mDNS is silently enabled by default in systemd-resolved. So if people enable Avahi now, how does that not cause problems?

In any case, I think that mDNS should be made opt-in, but doing so would break the setups of anyone who relies on the current default. I don't know how I would go about changing that.